node: metrics: add resource alignment metrics

ffromani opened the pull request:
/kind feature

#### What this PR does / why we need it:
Add metrics to give more feedback about aligned allocation of compute resources (xref: https://github.com/kubernetes/enhancements/pull/4810#discussion_r1738664727 https://github.com/kubernetes/enhancements/pull/4810#discussion_r1740964156)

#### Which issue(s) this PR fixes:
Fixes N/A

#### Special notes for your reviewer:
The current dimension, meant as possible valueset of labels, is currently 2 (two). We plan to add more possible alignments (e.g. https://github.com/kubernetes/enhancements/pull/4810) in the future, but this number is expected to grow both very slowly and very rarely. I'd expect the dimension to stay well below 10 (ten) in the very worst case.

#### Does this PR introduce a user-facing change?
```release-note
Added a kubelet metric `container_aligned_compute_resources_count`  to report the count of containers getting aligned compute resources
```

ffromani commented:
/sig node
/priority backlog

priority may be raised if turns to be useful (conversation ongoing) with the related KEP

ffromani commented:
/test pull-kubernetes-e2e-kind

ffromani commented:
/retest

ffromani commented:
> ok, as weird as it seems to me, will check locally before more retests

it seems it was actually all CI issues

ffromani commented:
> > ok, as weird as it seems to me, will check locally before more retests
> 
> it seems it was actually all CI issues

Maybe, but surely had a bug mismatching labels. Fixed in the latest upload.

bart0sh commented:
/triage accepted
/lgtm

bart0sh commented:
/assign @mrunalp @SergeyKanzhelev 
for approval

SergeyKanzhelev commented:
@ffromani can you please beef up the release note text. Something like: "Metric `container_aligned_compute_resources_count` with dimension `xx` was introduced that will be used to blah." It will make release team job easier closer to release time

SergeyKanzhelev reviewed (commented):
looks good overall. I like having a metric for this. I have hard time to reason how accurate it will be based on your comment in code. Do you have this understanding?

ffromani commented:
/hold

@SergeyKanzhelev thanks for the review, good points (relnote too terse and some concern about accuracy). I'll reword and clarify both so we can iterate.

ffromani commented:
@SergeyKanzhelev I think I addressed both your last points in my latest upload, PTAL

ffromani commented:
/hold

want to run a bit extra verification on my side

ffromani commented:
/retest-required
```
{Failed;  === RUN   TestTerminationOrderingObeysGrace
    kuberuntime_termination_order_test.go:278: expected delay for container sc3 = 1, got 0
--- FAIL: TestTerminationOrderingObeysGrace (3.00s)
```
seems very much unrelated

ffromani committed:
node: metrics: add resource alignment metrics

In order to improve the observability of the resource management
in kubelet, cpu allocation and NUMA alignment, we add more metrics
to report if resource alignment is in effect.

The more precise reporting would probably be using pod status,
but this would require more invasive and riskier changes,
and possibly extra interactions to the APIServer.

We start adding metrics to report if containers got their
compute resources aligned.
If metrics are growing, the assingment is working as expected;
If metrics stay consistent, perhaps at zero, no resource
alignment is done.

Extra fixes brought by this work
- retroactively add labels for existing tests
- running metrics test demands precision accounting to avoid flakes;
  ensure the node state is restored pristine between each test, to
  minimize the aforementioned risk of flakes.
- The test pod command line was wrong, with this the pod could not
  reach Running state. That gone unnoticed so far because
  no test using this utility function actually needed a pod
  in running state.

Signed-off-by: Francesco Romani <fromani@redhat.com>

haircommander commented:
/cc 
/triage accepted

haircommander commented on pkg/kubelet/cm/topologymanager/scope_pod.go:
```diff
@@ -62,6 +62,10 @@ func (s *podScope) Admit(pod *v1.Pod) lifecycle.PodAdmitResult {
 	}
+	if IsAlignmentGuaranteed(s.policy) {
+		// increment only if we know we allocate aligned resources.
+		metrics.ContainerAlignedComputeResources.WithLabelValues(metrics.AlignedNUMAZone).Inc()
```
should we separate pod and container alignment? do we expect a container to be aligned IFF a pod is aligned? if so, will the value of 2 per aligned container be confusing?

ffromani replied:
The semantic can be confusing but I'm struggling to make it easier to consume if one's not deep into details (which will likely bring in turn questions about the design of the features, but let's not digress now).

with scope=pod, the topology manager (TM) will place ALL the pod's containers on the same NUMA zone, failing to admit if not possible. So if we are here, it means all the containers are BOTH aligned AND all on the _same_ NUMA zone.

with scope=container, TM will place ALL the pod's containers on ANY available NUMA zone, failing to admit if any container can't get aligned resources. So, if we are here, it means ALL containers got NUMA aligned resources, but they can be scattered on different NUMA zones.

haircommander replied:
so I guess the question is: should the metric footprint look different for those two cases?  would an admin want to quicky tell what it meant by alignedNUMAZone? if they have distinct code paths and distinct results, it seems like something that would warrant an  additional label value

ffromani replied:
good point. I can add a metric label "scope" whose value can be "pod" or "container"

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/hold

need to adjust the tests

SergeyKanzhelev commented:
@ffromani please ping when it's time to review

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/retest

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/retest-required

ffromani commented:
@SergeyKanzhelev ready for review!

ffromani commented:
pull-kubernetes-node-kubelet-serial-crio-cgroupv1 seems to have failed for infra issues;on top of that, there's nothing cgroup specific, or even cgroup-related, in this PR.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1

ffromani commented:
serial-crio-cgroupv* lane failures are unrelated. Other lanes e.g. serial-cpu-manager exercise this flow correctly.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/hold cancel

SergeyKanzhelev commented:
> @SergeyKanzhelev ready for review!

Is it ready now? Can you please squash commits

ffromani commented:
> > @SergeyKanzhelev ready for review!
> 
> Is it ready now? Can you please squash commits

yes, but squashing commits anyway

ffromani commented:
@SergeyKanzhelev squashed!

ffromani commented:
after the squash, https://github.com/kubernetes/kubernetes/pull/127506 needs to be rebased on top of this one

ffromani commented:
gentle ping because we have quite some PRs depending on this one

SergeyKanzhelev commented on pkg/kubelet/cm/topologymanager/policy.go:
```diff
@@ -30,6 +30,14 @@ type Policy interface {
+// IsAlignmentGuaranteed return true if the given policy guarantees that either
+// the compute resources will be allocated within a NUMA boundary, or the allocation will fail at all.
+func IsAlignmentGuaranteed(p Policy) bool {
+	// TODO: should we add another field to the Policy interface to explicitly convey the alignment promise?
```
I have hard time judging if this TODO is fine to keep. If we do not plan to fix it any time soon, perhaps it can be rephrased into the explanation of the edge cases?

ffromani replied:
good point, rephrasing

SergeyKanzhelev commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -127,10 +127,21 @@ const (
+	AlignScopeContainer = "container"
+
+	AlignedPhysicalCPU = "physical_cpu"
+	AlignedNUMAZone    = "numa_zone"
```
do we refer to numa node as zone in other places? Is zone the same as node?

ffromani replied:
AFAIK "numa zones" "numa cells" and "numa nodes" are all synonyms, and I'd like to avoid the overinflaction of the "node" term.

SergeyKanzhelev replied:
Do we ever use NUMA zone term? I see only numa nodes for example here: https://kubernetes.io/docs/tasks/administer-cluster/memory-manager/ Will it be more confusing to mix terms interchangeably than we gain by not mixing "nodes" and 
"NUMA nodes"?

ffromani replied:
ok, changing to "NUMA nodes"

SergeyKanzhelev commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -917,6 +928,16 @@ var (
 		},
 	)
+
+	ContainerAlignedComputeResources = metrics.NewCounterVec(
```
Can you put it next to `CPUManagerPinningRequestsTotal` metric? They seems to be related. Also what is the difference between `NewCounterVec` and `NewCounter`?

ffromani replied:
sure, moving.
Regarding Counter vs CounterVec:
CounterVec bundles a set of Counters that all share the same Description. We use it because we want to count the same value partitioned by various dimensions: scope and boundary.

SergeyKanzhelev commented on pkg/kubelet/metrics/metrics.go:
```diff
@@ -917,6 +928,16 @@ var (
+		&metrics.CounterOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           ContainerAlignedComputeResourcesNameKey,
+			Help:           "Cumulative number of aligned compute resources allocated to containers by alignment type.",
```
are we accounting for a failed admissions somehow? Or only for successful?

ffromani replied:
yes, using `CPUManagerPinningErrorsTotal` and `TopologyManagerAdmissionErrorsTotal`

SergeyKanzhelev commented on test/e2e_node/cpu_manager_metrics_test.go:
```diff
@@ -91,11 +91,12 @@ var _ = SIGDescribe("CPU Manager Metrics", framework.WithSerial(), feature.CPUMa
 		})
 
-		ginkgo.It("should report zero pinning counters after a fresh restart", func(ctx context.Context) {
+		ginkgo.It("[pinning] should report zero pinning counters after a fresh restart", func(ctx context.Context) {
```
How do you plan to use the `[pinning]` tag?

ffromani replied:
It's the easiest and simplest way I know to run a subset of tests without using crazy regexes using `gingko --focus`

SergeyKanzhelev replied:
My question is whether you want to use it just locally or also in test-infra? We prefer to switch to the ginkgo tags, and we started moving there. Ginkgo tags make it easier to query which tests to run

SergeyKanzhelev replied:
If it is just for local debugging - I would question the value of it being added permanently. Is it easy to understand what pinning vs. alignment means? I have difficulty understanding it

ffromani replied:
I can move to labels if it helps

ffromani replied:
actually removed

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

SergeyKanzhelev commented:
/retest
